future_canceler.go

```diff
@@ -0,0 +1,17 @@
+package tarantool
+
+type futureCanceler interface {
+	Cancel(fut *Future, err error) error
+}
+
+type futCancelerImpl struct {
+	cancel func(fut *Future, err error) error
+}
+
+func newFutCancelerImpl(cancelFunc func(fut *Future, err error) error) *futCancelerImpl {
+	return &futCancelerImpl{cancel: cancelFunc}
+}
+
+func (fc *futCancelerImpl) Cancel(fut *Future, err error) error {
+	return fc.cancel(fut, err)
+}
```

Comment:
In such implementation it will be more clear to replace the interface by a closure. In a client code it will look

<details><summary>like this:</summary>

```patch
--- a/connection.go
+++ b/connection.go
@@ -140,8 +140,6 @@ type Connection struct {
        state   uint32
        dec     *msgpack.Decoder
        lenbuf  [PacketLengthBytes]byte
-
-       futCanceler futureCanceler
 }
 
 var _ = Connector(&Connection{}) // Check compatibility with connector interface.
@@ -337,7 +335,6 @@ func Connect(addr string, opts Opts) (conn *Connection, err error) {
                        return nil, err
                }
        }
-       conn.futCanceler = newFutCancelerImpl(conn.cancelFuture)
 
        return conn, err
 }
@@ -391,16 +388,6 @@ func (conn *Connection) Handle() interface{} {
        return conn.opts.Handle
 }
 
-// cancelFuture removes the passed future from the internal queue of Future objects.
-func (conn *Connection) cancelFuture(fut *Future, err error) error {
-       if fut == nil {
-               return fmt.Errorf("passed nil future")
-       }
-       fut.SetError(err)
-       conn.fetchFuture(fut.requestId)
-       return nil
-}
-
 func (conn *Connection) dial() (err error) {
        var connection net.Conn
        network := "tcp"
@@ -745,7 +732,14 @@ func (conn *Connection) reader(r *bufio.Reader, c net.Conn) {
 }
 
 func (conn *Connection) newFuture(ctx context.Context) (fut *Future) {
-       fut = NewFuture(ctx, conn.futCanceler)
+       fut = NewFuture(ctx, func(fut *Future, err error) error {
+               if fut == nil {
+                       return fmt.Errorf("passed nil future")
+               }
+               fut.SetError(err)
+               conn.fetchFuture(fut.requestId)
+               return nil
+       })
        if conn.rlimit != nil && conn.opts.RLimitAction == RLimitDrop {
                select {
                case conn.rlimit <- struct{}{}:
````
</details>

or

<details><summary>or like this:</summary>

```patch
--- a/connection.go
+++ b/connection.go
@@ -140,8 +140,6 @@ type Connection struct {
        state   uint32
        dec     *msgpack.Decoder
        lenbuf  [PacketLengthBytes]byte
-
-       futCanceler futureCanceler
 }
 
 var _ = Connector(&Connection{}) // Check compatibility with connector interface.
@@ -337,7 +335,6 @@ func Connect(addr string, opts Opts) (conn *Connection, err error) {
                        return nil, err
                }
        }
-       conn.futCanceler = newFutCancelerImpl(conn.cancelFuture)
 
        return conn, err
 }
@@ -391,16 +388,6 @@ func (conn *Connection) Handle() interface{} {
        return conn.opts.Handle
 }
 
-// cancelFuture removes the passed future from the internal queue of Future objects.
-func (conn *Connection) cancelFuture(fut *Future, err error) error {
-       if fut == nil {
-               return fmt.Errorf("passed nil future")
-       }
-       fut.SetError(err)
-       conn.fetchFuture(fut.requestId)
-       return nil
-}
-
 func (conn *Connection) dial() (err error) {
        var connection net.Conn
        network := "tcp"
@@ -744,8 +731,21 @@ func (conn *Connection) reader(r *bufio.Reader, c net.Conn) {
        }
 }
 
+type testCanceler struct {
+       conn *Connection
+}
+
+func (canceler *testCanceler) Cancel(fut *Future, err error) error {
+       if fut == nil {
+               return fmt.Errorf("passed nil future")
+       }
+       fut.SetError(err)
+       canceler.conn.fetchFuture(fut.requestId)
+       return nil
+}
+
 func (conn *Connection) newFuture(ctx context.Context) (fut *Future) {
-       fut = NewFuture(ctx, conn.futCanceler)
+       fut = NewFuture(ctx, &testCanceler{conn: conn})
        if conn.rlimit != nil && conn.opts.RLimitAction == RLimitDrop {
                select {
                case conn.rlimit <- struct{}{}:
```

</details>

**(you can still save the closure or the testCanceler object in `Connection.futureCanceler` to avoid allocations for each Future if you want)**
 

It's too complicated right now.

Reply:
Fixed.

Reply:
Fixed.